// @adonisjs/validator/build/providers/ValidatorProvider.js
"use strict";
/*
* @adonisjs/validator
*
* (c) Harminder Virk <virk@adonisjs.com>
*
* For the full copyright and license information, please view the LICENSE
* file that was distributed with this source code.
*/
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Rules_1 = require("../src/Rules");
const Schema_1 = require("../src/Schema");
const Validator_1 = require("../src/Validator");
const Request_1 = __importDefault(require("../src/Bindings/Request"));
/**
 * Provider to register validator with the IoC container
 */
class ValidationProvider {
    constructor(container) {
        this.container = container;
    }
    register() {
        this.container.singleton('Adonis/Core/Validator', () => {
            return {
                validator: Validator_1.validator,
                schema: Schema_1.schema,
                rules: Rules_1.rules,
            };
        });
    }
    boot() {
        this.container.with(['Adonis/Core/Request', 'Adonis/Core/Validator'], (Request, Validator) => {
            Request_1.default(Request, Validator.validator.validate);
        });
    }
}
exports.default = ValidationProvider;
