// @adonisjs/auth/build/src/AuthManager/index.js
"use strict";
/*
* @adonisjs/auth
*
* (c) Harminder Virk <virk@adonisjs.com>
*
* For the full copyright and license information, please view the LICENSE
* file that was distributed with this source code.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthManager = void 0;
const utils_1 = require("@poppinss/utils");
const Auth_1 = require("../Auth");
/**
 * Auth manager to manage guards and providers object. The extend API can
 * be used to add custom guards and providers
 */
class AuthManager {
    constructor(container, config) {
        this.container = container;
        this.config = config;
        /**
         * Extended set of providers
         */
        this.extendedProviders = new Map();
        /**
         * Extend set of guards
         */
        this.extendedGuards = new Map();
        /**
         * Reference to the default guard
         */
        this.defaultGuard = this.config.guard;
        const validator = new utils_1.ManagerConfigValidator(config, 'auth', 'config/auth');
        validator.validateDefault('guard');
        validator.validateList('list', 'guard');
    }
    /**
     * Verifies and returns an instance of the event emitter
     */
    getEmitter() {
        const hasEmitter = this.container.hasBinding('Adonis/Core/Event');
        if (!hasEmitter) {
            throw new utils_1.Exception('"Adonis/Core/Event" is required by the auth provider');
        }
        return this.container.use('Adonis/Core/Event');
    }
    /**
     * Lazily makes an instance of the lucid provider
     */
    makeLucidProvider(config) {
        return new (require('../UserProviders/Lucid').LucidProvider)(this.container, config);
    }
    /**
     * Lazily makes an instance of the database provider
     */
    makeDatabaseProvider(config) {
        const Database = this.container.use('Adonis/Lucid/Database');
        return new (require('../UserProviders/Database').DatabaseProvider)(this.container, config, Database);
    }
    /**
     * Returns an instance of the extended provider
     */
    makeExtendedProvider(config) {
        const providerCallback = this.extendedProviders.get(config.driver);
        if (!providerCallback) {
            throw new utils_1.Exception(`Invalid provider "${config.driver}"`);
        }
        return providerCallback(this.container, config);
    }
    /**
     * Lazily makes an instance of the token database provider
     */
    makeTokenDatabaseProvider(config) {
        const Database = this.container.use('Adonis/Lucid/Database');
        return new (require('../TokenProviders/Database').TokenDatabaseProvider)(config, Database);
    }
    /**
     * Returns an instance of the session guard
     */
    makeSessionGuard(mapping, config, provider, ctx) {
        const { SessionGuard } = require('../Guards/Session');
        return new SessionGuard(mapping, config, this.getEmitter(), provider, ctx);
    }
    /**
     * Returns an instance of the session guard
     */
    makeOatGuard(mapping, config, provider, ctx) {
        const { OATGuard } = require('../Guards/Oat');
        const tokenProvider = this.makeTokenProviderInstance(config.tokenProvider);
        return new OATGuard(mapping, config, this.getEmitter(), provider, ctx, tokenProvider);
    }
    /**
     * Returns an instance of the extended guard
     */
    makeExtendedGuard(mapping, config, provider, ctx) {
        const guardCallback = this.extendedGuards.get(config.driver);
        if (!guardCallback) {
            throw new utils_1.Exception(`Invalid guard driver "${config.driver}" property`);
        }
        return guardCallback(this.container, mapping, config, provider, ctx);
    }
    /**
     * Makes instance of a provider based upon the driver value
     */
    makeUserProviderInstance(providerConfig) {
        if (!providerConfig || !providerConfig.driver) {
            throw new utils_1.Exception('Invalid auth config, missing "provider" or "provider.driver" property');
        }
        switch (providerConfig.driver) {
            case 'lucid':
                return this.makeLucidProvider(providerConfig);
            case 'database':
                return this.makeDatabaseProvider(providerConfig);
            default:
                return this.makeExtendedProvider(providerConfig);
        }
    }
    /**
     * Makes instance of a provider based upon the driver value
     */
    makeTokenProviderInstance(providerConfig) {
        if (!providerConfig || !providerConfig.driver) {
            throw new utils_1.Exception('Invalid auth config, missing "tokenProvider" or "tokenProvider.driver" property');
        }
        switch (providerConfig.driver) {
            case 'database':
                return this.makeTokenDatabaseProvider(providerConfig);
            default:
                throw new utils_1.Exception(`Invalid token provider "${providerConfig.driver}"`);
        }
    }
    /**
     * Makes guard instance for the defined driver inside the
     * mapping config.
     */
    makeGuardInstance(mapping, mappingConfig, provider, ctx) {
        if (!mappingConfig || !mappingConfig.driver) {
            throw new utils_1.Exception('Invalid auth config, missing "driver" property');
        }
        switch (mappingConfig.driver) {
            case 'session':
                return this.makeSessionGuard(mapping, mappingConfig, provider, ctx);
            case 'oat':
                return this.makeOatGuard(mapping, mappingConfig, provider, ctx);
            default:
                return this.makeExtendedGuard(mapping, mappingConfig, provider, ctx);
        }
    }
    /**
     * Make an instance of a given mapping for the current HTTP request.
     */
    makeMapping(ctx, mapping) {
        const mappingConfig = this.config.list[mapping];
        const provider = this.makeUserProviderInstance(mappingConfig.provider);
        return this.makeGuardInstance(mapping, mappingConfig, provider, ctx);
    }
    /**
     * Returns an instance of the auth class for the current request
     */
    getAuthForRequest(ctx) {
        return new Auth_1.Auth(this, ctx);
    }
    extend(type, name, callback) {
        if (type === 'provider') {
            this.extendedProviders.set(name, callback);
        }
        if (type === 'guard') {
            this.extendedGuards.set(name, callback);
        }
    }
}
exports.AuthManager = AuthManager;
